sdk/core/Azure.Core/src/Shared/AsyncLockWithValue.cs

```diff
@@ -1,25 +1,63 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#nullable enable
+
 using System;
 using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
 using System.Threading;
 using System.Threading.Tasks;
 using Azure.Core.Pipeline;
 
-namespace Azure.Identity
+namespace Azure.Core
 {
     /// <summary>
     /// Primitive that combines async lock and value cache
     /// </summary>
     /// <typeparam name="T"></typeparam>
     internal sealed class AsyncLockWithValue<T>
     {
-        private readonly object _syncObj = new object();
-        private Queue<TaskCompletionSource<Lock>> _waiters;
+        private readonly object _syncObj = new();
+        private Queue<TaskCompletionSource<LockOrValue>>? _waiters;
         private bool _isLocked;
         private bool _hasValue;
-        private T _value;
+        private long _index;
+        private T? _value;
+
+        public bool HasValue
+        {
+            get
+            {
+                lock (_syncObj)
+                {
+                    return _hasValue;
+                }
+            }
+        }
+
+        public AsyncLockWithValue() { }
+
+        public AsyncLockWithValue(T value)
+        {
+            _hasValue = true;
+            _value = value;
+        }
+
+        public bool TryGetValue(out T? value)
+        {
+            lock (_syncObj)
+            {
+                if (_hasValue)
```

Comment:
Can the value ever be removed? If not, you might want to first check if the value is there (without locking) and then only when it's not there, lock. In such case, I think you also need to make the value field volatile such that it gets synchronized.

Reply:
Of course, I am not even sure it's worth the trouble. Depends on how hot this path it.

Reply:
Yeah, it's not a hot path where the type is used right now. I'd rather keep it simpler until we get any perf issues.